### PR TITLE
feat: add webhook receiver server

### LINKED ISSUE #578: Create a Webhook Receiver service
Acceptance criteria:
* [x] There is an application to start a webhook service (under `cmd/`).
* [x] The service listens on a a port specified as a command line argument.
* [x] There is an HTTP endpoint that can receive requests from GitHub (see #577).
* [x] The endpoint validates the incoming request with a secret (for now, predefined from an Environment Variable, later stored per repo or branch).

Notes:
* This is a bootstrap task with basic logic. The endpoint does not have to do anything with the payload other than validating it.
* No sub/unsub yet, that's not the Webhook Receiver's job. Now, it's a manual task to add webhooks.
* To test, you may need create a payload locally and send it with a script, or start ngrok (or similar service) to expose the endpoint, so GitHub can call it.

### PR DESCRIPTION
**Pull request by yitsushi**
* It can listen on a port set by a cli flag.
* It has a logger, similar logic as we have in other services (wgo oss, wge, progressive delivery, pipeline controller).
* It can receive webhook requests on `/callback?provider=github`. Not the root, case we need more endpoints, it will be easier to just and new ones instead of moveing existing part over or even send out a "breaking changes" notice.
* Validates the request. The secret key is coming from the `WEBHOOK_HMAC_KEY` environment variable.
* Parse Request with jenkins-x/go-scm[^1].
* Log some information about the event.
* Webhook Handler has tests. I tried to cover most cases.

At the end, we don't have to use and implement our own hmac validation[^2] as the go-scm[^1] library does that for us

Closes #578 [^3]

[^1]: http://github.com/jenkins-x/go-scm
[^2]: https://github.com/weaveworks/tf-controller/issues/577
[^3]: https://github.com/weaveworks/tf-controller/issues/578

### CODE REVIEW

**Comment by yitsushi** [comment]
```
❯ WEBHOOK_HMAC_KEY="xxxx" go run ./cmd/webhook-receiver --log-level debug
{"level":"info","ts":"2023-04-26T14:48:16.877+0200","msg":"Starting server","kind":"webhook","addr":"[::]:8080"}
{"level":"debug","ts":"2023-04-26T14:48:47.189+0200","msg":"incoming hmac requesr is valid","organization":"yitsushi","repository":"webhook-tester","kind":"issue_comment"}
```

**Comment by chanwit** [comment]
Thank you @yitsushi 

Overall, it looks good to me. You also used `logr` and Flux's log packages to handle logging, which is great. 
Please continue with the next steps.

**Comment by yitsushi** [comment]
Now it's ready for a full review.

File: internal/server/webhook/callback_handler_test.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+			responseRecorder := httptest.NewRecorder()
+
+			handler.ServeHTTP(responseRecorder, request)
+			assert.Equal(t, http.StatusMethodNotAllowed, responseRecorder.Code)
+		})
+	}
+}
+
+func TestCallbackHandler_InvalidProvider(t *testing.T) {
+	testCases := []struct {
```
Comment (chanwit) [inline internal/server/webhook/callback_handler_test.go:42]:
```suggestion
	g := NewGomegaWithT(t)

	testCases := []struct {
```

File: internal/server/webhook/callback_handler_test.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+		{
+			name:        "invalid provider",
+			url:         "/callback?provider=something",
+			code:        http.StatusBadRequest,
+			logContains: "Unsupported GIT_KIND value: something",
+		},
+	}
+
+	var buf bytes.Buffer
+	var log logr.Logger = buflogr.NewWithBuffer(&buf)
```
Comment (chanwit) [inline internal/server/webhook/callback_handler_test.go:63]:
```suggestion
	var (
		buf bytes.Buffer
		log = buflogr.NewWithBuffer(&buf)
	)
```

File: internal/server/webhook/callback_handler_test.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			request := httptest.NewRequest(http.MethodPost, testCase.url, bytes.NewReader(body))
+			request.Header = header
+			responseRecorder := httptest.NewRecorder()
+
+			handler.ServeHTTP(responseRecorder, request)
+			assert.Equal(t, testCase.code, responseRecorder.Code)
+			assert.Contains(t, buf.String(), testCase.logContains)
```
Comment (chanwit) [inline internal/server/webhook/callback_handler_test.go:79]:
```suggestion
			g.Expect(responseRecorder.Code).To(Equal(testCase.code))
			g.Expect(buf.String()).To(ContainSubstring(testCase.logContains))
```

File: internal/server/webhook/callback_handler_test.go [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+		})
+	}
+}
+
+func TestCallbackHandler_MissingHMAC(t *testing.T) {
+	os.Setenv("WEBHOOK_HMAC_KEY", webhookSecret)
```
Comment (chanwit) [inline internal/server/webhook/callback_handler_test.go:85]:
```suggestion
	g := NewGomegaWithT(t)

	os.Setenv("WEBHOOK_HMAC_KEY", webhookSecret)
```

File: internal/server/webhook/callback_handler_test.go [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+	var buf bytes.Buffer
+	var log logr.Logger = buflogr.NewWithBuffer(&buf)
```
Comment (chanwit) [inline internal/server/webhook/callback_handler_test.go:88]:
```suggestion
	var (
		buf bytes.Buffer
		log = buflogr.NewWithBuffer(&buf)
	)
```

File: internal/server/webhook/callback_handler_test.go [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+
+	handler := webhook.NewCallbackHandler(log)
+	body, header, err := readRequestFixture("pull_request_event")
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (chanwit) [inline internal/server/webhook/callback_handler_test.go:94]:
```suggestion
	g.Expect(err).NotTo(HaveOccurred())
```

File: internal/server/webhook/callback_handler_test.go [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+	header.Del(webhook.HubSignatureSHA256Header)
+	header.Del(webhook.HubSignatureSHA1Header)
+
+	request := httptest.NewRequest(http.MethodPost, "/callback?provider=github", bytes.NewReader(body))
+	request.Header = header
+	responseRecorder := httptest.NewRecorder()
+
+	handler.ServeHTTP(responseRecorder, request)
+	assert.Equal(t, http.StatusUnauthorized, responseRecorder.Code)
+	assert.Contains(t, buf.String(), "missing hmac signature")
```
Comment (chanwit) [inline internal/server/webhook/callback_handler_test.go:105]:
```suggestion
	g.Expect(responseRecorder.Code).To(Equal(http.StatusUnauthorized))
	g.Expect(buf.String()).To(ContainSubstring("missing hmac signature"))
```

File: internal/server/webhook/callback_handler_test.go [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+func TestCallbackHandler_InvalidHMAC(t *testing.T) {
+	os.Setenv("WEBHOOK_HMAC_KEY", "invalid-key")
```
Comment (chanwit) [inline internal/server/webhook/callback_handler_test.go:109]:
```suggestion
	g := NewGomegaWithT(t)

	os.Setenv("WEBHOOK_HMAC_KEY", "invalid-key")
```

File: internal/server/webhook/callback_handler_test.go [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+
+	var buf bytes.Buffer
+	var log logr.Logger = buflogr.NewWithBuffer(&buf)
```
Comment (chanwit) [inline internal/server/webhook/callback_handler_test.go:112]:
```suggestion
	var (
		buf bytes.Buffer
		log = buflogr.NewWithBuffer(&buf)
	)
```

File: internal/server/webhook/callback_handler_test.go [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+
+	handler := webhook.NewCallbackHandler(log)
+	body, header, err := readRequestFixture("pull_request_event")
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (chanwit) [inline internal/server/webhook/callback_handler_test.go:118]:
```suggestion
	g.Expect(err).NotTo(HaveOccurred())
```

File: internal/server/webhook/callback_handler_test.go [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+
+	request := httptest.NewRequest(http.MethodPost, "/callback?provider=github", bytes.NewReader(body))
+	request.Header = header
+	responseRecorder := httptest.NewRecorder()
+
+	handler.ServeHTTP(responseRecorder, request)
+	assert.Equal(t, http.StatusBadRequest, responseRecorder.Code)
+	assert.Contains(t, buf.String(), "invalid webhook signature parsing webhook request")
```
Comment (chanwit) [inline internal/server/webhook/callback_handler_test.go:126]:
```suggestion
	g.Expect(responseRecorder.Code).To(Equal(http.StatusBadRequest))
	g.Expect(buf.String()).To(ContainSubstring("invalid webhook signature parsing webhook request"))
```

File: internal/server/webhook/callback_handler_test.go [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+func TestCallbackHandler_PullRequestEvent(t *testing.T) {
+	os.Setenv("WEBHOOK_HMAC_KEY", webhookSecret)
```
Comment (chanwit) [inline internal/server/webhook/callback_handler_test.go:130]:
```suggestion
	g := NewGomegaWithT(t)

	os.Setenv("WEBHOOK_HMAC_KEY", webhookSecret)
```

File: internal/server/webhook/callback_handler_test.go [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+
+	var buf bytes.Buffer
+	var log logr.Logger = buflogr.NewWithBuffer(&buf)
```
Comment (chanwit) [inline internal/server/webhook/callback_handler_test.go:133]:
```suggestion
	var (
		buf bytes.Buffer
		log = buflogr.NewWithBuffer(&buf)
	)
```

File: internal/server/webhook/callback_handler_test.go [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+
+	handler := webhook.NewCallbackHandler(log)
+	body, header, err := readRequestFixture("pull_request_event")
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (chanwit) [inline internal/server/webhook/callback_handler_test.go:139]:
```suggestion
	g.Expect(err).NotTo(HaveOccurred())
```

File: internal/server/webhook/callback_handler_test.go [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+
+	request := httptest.NewRequest(http.MethodPost, "/callback?provider=github", bytes.NewReader(body))
+	request.Header = header
+	responseRecorder := httptest.NewRecorder()
+
+	handler.ServeHTTP(responseRecorder, request)
+	assert.Equal(t, http.StatusAccepted, responseRecorder.Code)
+	assert.Contains(t, buf.String(), "incoming hmac request is valid kind pull_request")
```
Comment (chanwit) [inline internal/server/webhook/callback_handler_test.go:147]:
```suggestion
	g.Expect(responseRecorder.Code).To(Equal(http.StatusAccepted))
	g.Expect(buf.String()).To(ContainSubstring("incoming hmac request is valid kind pull_request"))
```

File: internal/server/webhook/callback_handler_test.go [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+func TestCallbackHandler_CommentEvent(t *testing.T) {
+	os.Setenv("WEBHOOK_HMAC_KEY", webhookSecret)
```
Comment (chanwit) [inline internal/server/webhook/callback_handler_test.go:151]:
```suggestion
	g := NewGomegaWithT(t)

	os.Setenv("WEBHOOK_HMAC_KEY", webhookSecret)
```

File: internal/server/webhook/callback_handler_test.go [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+
+	var buf bytes.Buffer
+	var log logr.Logger = buflogr.NewWithBuffer(&buf)
```
Comment (chanwit) [inline internal/server/webhook/callback_handler_test.go:154]:
```suggestion
	var (
		buf bytes.Buffer
		log = buflogr.NewWithBuffer(&buf)
	)
```

File: internal/server/webhook/callback_handler_test.go [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+
+	handler := webhook.NewCallbackHandler(log)
+	body, header, err := readRequestFixture("comment_event")
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (chanwit) [inline internal/server/webhook/callback_handler_test.go:160]:
```suggestion
	g.Expect(err).NotTo(HaveOccurred())
```

File: internal/server/webhook/callback_handler_test.go [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+
+	request := httptest.NewRequest(http.MethodPost, "/callback?provider=github", bytes.NewReader(body))
+	request.Header = header
+	responseRecorder := httptest.NewRecorder()
+
+	handler.ServeHTTP(responseRecorder, request)
+	assert.Equal(t, http.StatusAccepted, responseRecorder.Code)
+	assert.Contains(t, buf.String(), "incoming hmac request is valid kind issue_comment")
```
Comment (chanwit) [inline internal/server/webhook/callback_handler_test.go:168]:
```suggestion
	g.Expect(responseRecorder.Code).To(Equal(http.StatusAccepted))
	g.Expect(buf.String()).To(ContainSubstring("incoming hmac request is valid kind issue_comment"))
```

**Review by chanwit — CHANGES_REQUESTED** [review]
Thank you for writing a comprehensive set of tests @yitsushi :1st_place_medal:

**Comment by yitsushi** [comment]
Thank you, for the review, updated the tests to use `gomega`. Did not apply comments one by one, because:
- It was easier to just replace it in vim than pressing the "add to batch" button on all comments.
- I had a slight modification on suggestion, so I would have to make a new commit anyway.

**Comment by chanwit** [comment]
No worries @yitsushi :-)

**Review by chanwit — APPROVED** [review]
LGTM thank you @yitsushi :1st_place_medal:

**Comment by chanwit** [comment]
ping @yiannistri

**Comment by chanwit** [comment]
I'll go ahead merging this so we can continue on the next thing.

**Pull request merged**